### PR TITLE
Add SqueezeWave

### PR DESCRIPTION
**Pull request by vliu15**

### CODE REVIEW

**Comment by okuchaiev** [comment]
can you please resolve merge conflicts?

File: docs/source/tts/models.rst [hunk 1]
```diff
@@ -0,0 +1,40 @@
+Models
```
Comment (blisc) [inline docs/source/tts/models.rst:1]: Thank you for starting the TTS docs

File: Jenkinsfile [hunk 2]
```diff
@@ -662,6 +662,19 @@ pipeline {
             model.validation_ds.batch_size=4'
           }
         }
+        stage('SqueezeWave') {
```
Comment (blisc) [inline Jenkinsfile:665]: Let's figure out why this isn't being run

File: docs/source/tts/models.rst [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+======
+
+Currently, NeMo's TTS collection supports the following models:
+
+.. _WaveGlow_Model:
+
+WaveGlow
+--------
+
+WaveGlow :cite:`tts-models-prenger2018waveglow` is a Flow-based generative model that generates audio from mel spectrogram.
```
Comment (blisc) [inline docs/source/tts/models.rst:11]: `mel spectrograms.`

File: docs/source/tts/models.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+Comprised of several flow steps, WaveGlow learns an invertible mapping from a simple latent space to audio waveforms.
+
+    .. image:: waveglow.png
+        :align: center
+        :alt: waveglow model
+
+WaveGlow can be instantiated using :class:`WaveGlowModel<nemo.collections.tts.models.WaveGlowModel>` class.
```
Comment (blisc) [inline docs/source/tts/models.rst:18]: `using the`

File: docs/source/tts/models.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+-----------
+
+SqueezeWave :cite:`tts-models-zhai2020squeezewave` is a version of WaveGlow :cite:`tts-models-prenger2018waveglow` that simplifies the architecture of the WaveNet (WN) module by introducing depthwise separable convolutions and removing dual channels.
+SqueezeWave also uses larger group sizes, which reduces computation along the temporal axis and allows for less upsampling layers for mel spectrogram.
+
+    .. image:: squeezewave_wn.png
+        :align: center
+        :alt: squeezewave vs waveglow wavenet modules
+
+SqueezeWave can be instantiated using :class:`SqueezeWaveModel<nemo.collections.tts.models.SqueezeWaveModel>` class.
```
Comment (blisc) [inline docs/source/tts/models.rst:31]: `using the`

File: examples/tts/conf/squeezewave.yaml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    dataset:
+      cls: "nemo.collections.tts.data.datalayers.AudioDataset"
+      params:
+        manifest_filepath: ${validation_datasets}
+        max_duration: null
+        min_duration: 0.1
+        n_segments: -1
+        trim: false
+        # This value is hard-coded to reflect n_window_stride * lcm(2, pad_to)
+        truncate_to: 512
```
Comment (blisc) [inline examples/tts/conf/squeezewave.yaml:39]: Jason TODO

File: examples/tts/notebooks/SqueezeWave_TTS.ipynb [hunk 7]
```diff
@@ -0,0 +1,188 @@
+{
```
Comment (blisc) [inline examples/tts/notebooks/SqueezeWave_TTS.ipynb:1]: Can you add SqueezeWave to the current TTS tutorial instead of adding a new notebook?

File: nemo/collections/asr/parts/features.py [hunk 8]
```diff
@@ -146,6 +151,50 @@ def forward(self, input_data):
...
 
+# Create helper class for STFT that yields num_frames = num_samples / hop_length
+class STFTExactPad(STFT):
+    """adapted from Prem Seetharaman's https://github.com/pseeth/pytorch-stft"""
+
+    def __init__(self, *params, **kw_params):
+        super(STFTExactPad, self).__init__(*params, **kw_params)
+        self.pad_amount = (self.filter_length - self.hop_length) // 2
+
+    def inverse(self, magnitude, phase):
```
Comment (blisc) [inline nemo/collections/asr/parts/features.py:162]: When do you need to use this inverse?

File: nemo/collections/asr/parts/features.py [hunk 9]
```diff
@@ -314,7 +355,9 @@ def forward(self, x, seq_len):
             x = x.pow(self.mag_power)
 
         # dot with filterbank energies
-        x = torch.matmul(self.fb.to(x.dtype), x)
+        # disable autocast to get full range of mel values
+        with torch.cuda.amp.autocast(enabled=False):
+            x = torch.matmul(self.fb.to(x.dtype), x)
```
Comment (blisc) [inline nemo/collections/asr/parts/features.py:360]: Is this part necessary?

File: nemo/collections/asr/parts/segment.py [hunk 10]
```diff
@@ -133,7 +133,7 @@ def from_file(
         return cls(samples, sample_rate, target_sr=target_sr, trim=trim)
 
     @classmethod
-    def segment_from_file(cls, audio_file, target_sr=None, n_segments=0, trim=False):
+    def segment_from_file(cls, audio_file, target_sr=None, n_segments=0, trim=False, truncate_to=1):
```
Comment (blisc) [inline nemo/collections/asr/parts/segment.py:136]: Don't think you need to add `truncate_to` here

File: nemo/collections/tts/data/datalayers.py [hunk 11]
```diff
@@ -113,11 +119,96 @@ def __getitem__(self, index):
...
+        self.truncate_to = truncate_to
+
+        self.samples = []
+        for index in range(len(collection)):
+            example = collection[index]
+            with sf.SoundFile(example.audio_file, 'r') as f:
+                samples = f.read(dtype='float32').transpose()
+                self.samples.append(samples)
+        self.samples = np.concatenate(self.samples, axis=0)
+        self.samples = self.samples[: self.samples.shape[0] - (self.samples.shape[0] % self.n_segments), ...]
```
Comment (blisc) [inline nemo/collections/tts/data/datalayers.py:201]: Doesn't this approach "stitch" (for lack of a better words) two distinct audio samples together?

Comment (blisc) [reply nemo/collections/tts/data/datalayers.py:201]: Does this improve SqueezeWave's performance?

File: nemo/collections/tts/helpers/helpers.py [hunk 12]
```diff
@@ -204,7 +204,11 @@ def remove_weightnorm(model):
     waveglow = model
     for wavenet in waveglow.wavenet:
         wavenet.start = torch.nn.utils.remove_weight_norm(wavenet.start)
-        wavenet.in_layers = remove(wavenet.in_layers)
+        # SqueezeWave doesn't have weight norm in in_layers
+        try:
+            wavenet.in_layers = remove(wavenet.in_layers)
+        except ValueError:
+            pass
```
Comment (blisc) [inline nemo/collections/tts/helpers/helpers.py:211]: I think a better check would be `if instance(model, SqueezeWaveModel)` as opposed to try, except

File: nemo/collections/tts/models/squeezewave.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        self._validation_dl = self.__setup_dataloader_from_config(cfg, shuffle_should_be=False, name="validation")
+
+    @classmethod
+    def list_available_models(cls) -> 'Optional[Dict[str, str]]':
+        """TODO: Implement me!"""
+        pass
+
+    def export(self):
+        """TODO: Implement me!"""
+        pass
```
Comment (blisc) [inline nemo/collections/tts/models/squeezewave.py:212]: I think this method is no longer abstract and can be removed.

File: nemo/collections/tts/models/squeezewave.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        return torch.utils.data.DataLoader(dataset, collate_fn=dataset.collate_fn, **cfg.dataloader_params)
+
+    def setup_training_data(self, cfg):
+        self._train_dl = self.__setup_dataloader_from_config(cfg)
+
+    def setup_validation_data(self, cfg):
+        self._validation_dl = self.__setup_dataloader_from_config(cfg, shuffle_should_be=False, name="validation")
+
+    @classmethod
+    def list_available_models(cls) -> 'Optional[Dict[str, str]]':
```
Comment (blisc) [inline nemo/collections/tts/models/squeezewave.py:206]: Please give a checkpoint to @okuchaiev, so we can populate this method

File: nemo/collections/asr/parts/features.py (hunk 8 above)
Comment (blisc) [reply nemo/collections/asr/parts/features.py:162]: I think it's used in the denoiser. Can we remove `stft_exact_pad` parameters from the preprocessors then?

File: nemo/collections/asr/modules/audio_preprocessing.py [hunk 15]
```diff
@@ -199,6 +203,7 @@ def __init__(
         dither=1e-5,
         pad_to=16,
         frame_splicing=1,
+        stft_exact_pad=False,
```
Comment (blisc) [inline nemo/collections/asr/modules/audio_preprocessing.py:206]: Can we remove `stft_exact_pad` here?

File: nemo/collections/asr/parts/features.py [hunk 16]
```diff
@@ -170,6 +219,7 @@ def __init__(
         pad_to=16,
         max_duration=16.7,
         frame_splicing=1,
+        stft_exact_pad=False,
```
Comment (blisc) [inline nemo/collections/asr/parts/features.py:222]: Can we remove `stft_exact_pad` here?

File: nemo/collections/tts/modules/squeezewave.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+class OperationMode(Enum):
+    """Training or Inference (Evaluation) mode"""
+
+    training = 0
+    validation = 1
+    infer = 2
+
+
+@experimental  # TODO: Implement save_to() and restore_from()
```
Comment (blisc) [inline nemo/collections/tts/modules/squeezewave.py:64]: You can drop experimental

File: nemo/collections/tts/modules/squeezewave.py [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+class SqueezeWaveModule(NeuralModule, Exportable):
```
Comment (blisc) [inline nemo/collections/tts/modules/squeezewave.py:65]: This should only inherit from NeuralModule

File: nemo/collections/asr/modules/audio_preprocessing.py (hunk 15 above)
Comment (vliu15) [reply nemo/collections/asr/modules/audio_preprocessing.py:206]: SqueezeWave needs slightly different padding than the default STFT modules that we have

File: nemo/collections/asr/parts/features.py (hunk 8 above)
Comment (vliu15) [reply nemo/collections/asr/parts/features.py:161]: Inverse is needed for denoiser. `stft_exact_pad` is needed to know when to call `STFTExactPad`, which uses slightly different padding than `STFTPatch`

File: nemo/collections/asr/parts/features.py (hunk 9 above)
Comment (vliu15) [reply nemo/collections/asr/parts/features.py:359]: I don't think so. Remove?

File: nemo/collections/tts/data/datalayers.py (hunk 11 above)
Comment (vliu15) [reply nemo/collections/tts/data/datalayers.py:199]: I don't think this affects performance much. All this does is use more of the training data, since originally we are just sampling a (small) part of each wave file.

File: nemo/collections/tts/helpers/helpers.py (hunk 12 above)
Comment (vliu15) [reply nemo/collections/tts/helpers/helpers.py:211]: i think this may cause circular import problems

File: nemo/collections/asr/parts/features.py (hunk 16 above)
Comment (vliu15) [reply nemo/collections/asr/parts/features.py:221]: See above comment

File: nemo/collections/tts/models/squeezewave.py (hunk 14 above)
Comment (okuchaiev) [reply nemo/collections/tts/models/squeezewave.py:206]: yes, please. You can produce ".nemo" file by calling .save_to(...) on a trained model

File: nemo/collections/tts/helpers/helpers.py (hunk 12 above)
Comment (blisc) [reply nemo/collections/tts/helpers/helpers.py:211]: python will handle circular imports

File: Jenkinsfile (hunk 2 above)
Comment (blisc) [reply Jenkinsfile]: We can figure this out in a future PR

File: nemo/collections/tts/data/datalayers.py [hunk 19]
```diff
@@ -113,11 +117,94 @@ def __getitem__(self, index):
...
+        max_duration: Optional[float] = None,
+        min_duration: Optional[float] = None,
+        trim: Optional[bool] = False,
+        truncate_to: Optional[int] = 1,
+    ):
+        """
+        Mostly compliant with nemo.collections.asr.data.datalayers.AudioToTextDataset except it only returns Audio
+        without text. Dataset that loads tensors via a json file containing paths to audio files, transcripts, and
+        durations (in seconds). Each new line is a different sample. Note that text is required, but is ignored for
+        AudioDataset. Example below:
```
Comment (blisc) [inline nemo/collections/tts/data/datalayers.py:158]: Can you update this docstring so it describes what this dataset does?

File: tutorials/tts/2_TTS_test.ipynb [hunk 20]
```diff
@@ -0,0 +1,188 @@
+{
```
Comment (blisc) [inline tutorials/tts/2_TTS_test.ipynb:1]: Can you delete this notebook?

**Review by blisc — APPROVED** [review]
